cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,229 @@
+        """Log additional metadata corresponding to the currently active transformer stage.
+
+        Args:
+            *args: The additional metadata to log.
+            level: Logging level to control the amount of metadata that gets put into the context.
+        """
+
+    @abc.abstractmethod
+    def register_final(self, circuit: CIRCUIT_TYPE, transformer_name: str) -> None:
+        """Register the end of the currently active transformer stage.
+
+        Args:
+            circuit: Final transformed output circuit from the transformer stage.
+            transformer_name: Name of the (currently active) transformer stage which ends.
+        """
+
+
+@dataclasses.dataclass()
+class TransformerContext:
+    """Stores common configurable options for transformers."""
+
+    logger: Optional[TransformerStatsLoggerBase] = None
+    ignore_tags: Tuple[Hashable, ...] = ()
+    max_parallelism: int = 1
+
+
+TRANSFORMER_TYPE = Callable[[CIRCUIT_TYPE, TransformerContext], CIRCUIT_TYPE]
+TRANSFORMER_TYPE_T = TypeVar(
+    'TRANSFORMER_TYPE_T',
+    TRANSFORMER_TYPE[circuit.Circuit],
+    TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit],
+)
+TRANSFORMER_CLS_TYPE = TypeVar(
+    'TRANSFORMER_CLS_TYPE',
+    Type[TRANSFORMER_TYPE[frozen_circuit.FrozenCircuit]],
+    Type[TRANSFORMER_TYPE[circuit.Circuit]],
+)
+
+
+def _transformer_with_logging(
+    func: TRANSFORMER_TYPE[CIRCUIT_TYPE],
+    transformer_name: str,
+    circuit: CIRCUIT_TYPE,
+    context: TransformerContext,
+) -> CIRCUIT_TYPE:
+    """Helper to append logging functionality to transformers."""
+
+    if context.logger is not None:
+        context.logger.register_initial(circuit, transformer_name)
+    circuit = func(circuit, context)
+    if context.logger is not None:
+        context.logger.register_final(circuit, transformer_name)
+    return circuit
+
+
+def transformer_method(func: TRANSFORMER_TYPE[CIRCUIT_TYPE]) -> TRANSFORMER_TYPE[CIRCUIT_TYPE]:
+    """Decorator to verify API and append logging functionality to callable transformer methods.
+
+    For example:
+    >>> @transformer_method
+    >>> def convert_to_cz(circuit: cirq.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
+    >>>    ...
+
+    Args:
+        func: Transformer implemented as a method from (CIRCUIT_TYPE, Context) --> CIRCUIT_TYPE.
+
+    Returns:
+        Method corresponding to original transformers + additional logging boilerplate.
+    """
+
+    @functools.wraps(func)
+    def transformer_with_logging_func(
+        circuit: CIRCUIT_TYPE, context: TransformerContext
+    ) -> CIRCUIT_TYPE:
+        return _transformer_with_logging(func, func.__name__, circuit, context)
+
+    return cast(TRANSFORMER_TYPE[CIRCUIT_TYPE], transformer_with_logging_func)
+
+
+def transformer_class(cls: TRANSFORMER_CLS_TYPE) -> TRANSFORMER_CLS_TYPE:
+    """Decorator to verify API and append logging functionality to callable transformer class.
```

Comment:
Instead of a decorator, can we do this with a base class that has an `__init_subclass__` method? Something like this:
```python
class Transformer(abs.ABC):
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        # Note: would have to modify `transformer_method` to accept an optional `name` to use instead of `func.__name__`
        cls.__call__ = transformer_method(cls.__call__, name=cls.__name__)

class ConvertToSqrtISwaps(Transformer):
    def __call__(self, circuit.Circuit, context: cirq.TransformerContext) -> cirq.Circuit:
        ...
```
I find that class decorators can be rather brittle and `__init_subclass__` handles most cases.